server/interpreter.js

```diff
@@ -1393,39 +1430,131 @@ Interpreter.Object.prototype.valueOf = function() {
 
 /**
  * Create a new data object.
- * @param {Interpreter.Object} constructor Parent constructor function,
- *     or null if scope object.
+ * @param {Interpreter.Object=} proto Prototype object (or null);
+ *     defaults to this.OBJECT.
  * @return {!Interpreter.Object} New data object.
  */
-Interpreter.prototype.createObject = function(constructor) {
-  return this.createObjectProto(constructor &&
-                                constructor.properties['prototype']);
+Interpreter.prototype.createObject = function(proto) {
+  var p = (proto === undefined ? this.OBJECT : proto);
+  return new Interpreter.Object(p);
 };
 
 /**
- * Create a new data object.
+ * Class for a function
  * @param {Interpreter.Object} proto Prototype object.
+ * @constructor
+ */
+Interpreter.Function = function(proto) {
+  Interpreter.Object.call(this, proto);
+};
+
+Interpreter.Function.prototype = Object.create(Interpreter.Object.prototype);
+Interpreter.Function.prototype.constructor = Interpreter.Function;
+Interpreter.Function.prototype.class = 'Function';
+
+/**
+ * Add a prototype property to this function object, setting
+ * this.properties[prototype] to prototype and
+ * prototype.properites[constructor] to func.  If prototype is not
+ * specified, a newly-created object will be used instead.
+ * @param {!Interpreter} interpreter Interpreter to which this is attached.
+ * @param {Interpreter.Object=} prototype Prototype to add to this.
+ */
+Interpreter.Function.prototype.addPrototype = function(thisInterpreter, prototype) {
+  if (this.illegalConstructor) {
+    throw TypeError("You said this wasn't constructor!");
+  }
+  var protoObj = prototype ||
+      thisInterpreter.createObject();
```

Comment:
There was a `thisInterpreter.OBJECTPROTO` between the parens at some point, and search-and-replace didn't un-wrap it for me.

I miss `gofmt`.